packages/protocol/contracts/governance/Governance.sol

```diff
@@ -615,30 +653,66 @@ contract Governance is
    * @notice Approves the hash of a hotfix transaction(s).
    * @param hash The abi encoded keccak256 hash of the hotfix transaction(s) to be approved.
    */
-  function approveHotfix(bytes32 hash) external hotfixNotExecuted(hash) onlyApprover {
-    hotfixes[hash].approved = true;
-    emit HotfixApproved(hash);
+  function approveHotfix(
+    bytes32 hash
+  ) external hotfixNotExecuted(hash) onlyApproverOrSecurityCouncil {
+    if (msg.sender == approver) {
+      hotfixes[hash].approved = true;
+    } else {
+      if (isL2()) {
+        hotfixes[hash].councilApproved = true;
+      } else {
+        revert("Hotfix approval by security council is not available on L1.");
+      }
+    }
+    emit HotfixApproved(hash, msg.sender);
   }
 
   /**
    * @notice Whitelists the hash of a hotfix transaction(s).
    * @param hash The abi encoded keccak256 hash of the hotfix transaction(s) to be whitelisted.
    */
-  function whitelistHotfix(bytes32 hash) external hotfixNotExecuted(hash) {
-    hotfixes[hash].whitelisted[msg.sender] = true;
+  function whitelistHotfix(bytes32 hash) external hotfixNotExecuted(hash) onlyL1 {
+    hotfixes[hash].deprecated_whitelisted[msg.sender] = true;
     emit HotfixWhitelisted(hash, msg.sender);
   }
 
   /**
-   * @notice Gives hotfix a prepared epoch for execution.
+   * @notice Gives hotfix a time limit for execution.
    * @param hash The hash of the hotfix to be prepared.
    */
   function prepareHotfix(bytes32 hash) external hotfixNotExecuted(hash) {
-    require(isHotfixPassing(hash), "hotfix not whitelisted by 2f+1 validators");
-    uint256 epoch = getEpochNumber();
-    require(hotfixes[hash].preparedEpoch < epoch, "hotfix already prepared for this epoch");
-    hotfixes[hash].preparedEpoch = epoch;
-    emit HotfixPrepared(hash, epoch);
+    if (isL2()) {
+      uint256 _currentTime = now;
+      require(hotfixExecutionTimeWindow > 0, "Hotfix execution time window not set");
+      require(
+        hotfixes[hash].executionTimeLimit < _currentTime,
```

Comment:
Since we are reading hotfixes[hash] we should extract it to local storage variable to save some gas